src/victory-util/add-events.js

```diff
@@ -14,30 +46,86 @@ export default (WrappedComponent) => {
+    shouldComponentUpdate(nextProps) {
+      const calculatedValues = this.getCalculatedValues(nextProps);
+
+      // re-render without additional checks when component is animated
+      if (this.props.animate || this.props.animating) {
+        this.cacheValues(calculatedValues);
+        return true;
       }
-      this.setupEvents(newProps);
+
+      // check for any state changes triggered by events or shared events
+      const calculatedState = this.getStateChanges(nextProps, calculatedValues);
+      if (!areVictoryPropsEqual(this.calculatedState, calculatedState)) {
+        this.cacheValues(calculatedValues);
+        this.calculatedState = calculatedState;
+        return true;
+      }
+
+      // check whether props have changed
+      if (!areVictoryPropsEqual(this.props, nextProps)) {
+        this.cacheValues(calculatedValues);
+        return true;
+      }
+
+      return false;
     }
 
-    setupEvents(props) {
+    // compile all state changes from own and parent state. Order doesn't matter, as any state
+    // state change should trigger a re-render
+    getStateChanges(props, calculatedValues) {
+      const { hasEvents, getSharedEventState } = calculatedValues;
+      if (!hasEvents) { return {}; }
+
+      const getState = (key, type) => {
+        const result = defaults({}, this.getEventState(key, type), getSharedEventState(key, type));
+        return isEmpty(result) ? undefined : result;
+      };
+
+      options = options || {};
+      const components = options.components || defaultComponents;
+      return components.map((component) => {
+        if (!props.standalone && component.name === "parent") {
+           // don't check for changes on parent props for non-standalone components
+          return undefined;
+        } else {
+          return component.index !== undefined ?
+          getState(component.index, component.name) :
+          calculatedValues.dataKeys.map((key) => getState(key, component.name));
+        }
+      }).filter(Boolean);
+    }
+
+    getCalculatedValues(props) {
       const { sharedEvents } = props;
       const components = WrappedComponent.expectedComponents;
-      this.componentEvents = Events.getComponentEvents(props, components);
-      this.getSharedEventState = sharedEvents && isFunction(sharedEvents.getEventState) ?
+      const componentEvents = Events.getComponentEvents(props, components);
+      const getSharedEventState = sharedEvents && isFunction(sharedEvents.getEventState) ?
         sharedEvents.getEventState : () => undefined;
-      this.baseProps = this.getBaseProps(props);
-      this.dataKeys = Object.keys(this.baseProps).filter((key) => key !== "parent");
-      this.hasEvents = props.events || props.sharedEvents || this.componentEvents;
-      this.events = this.getAllEvents(props);
+      const baseProps = this.getBaseProps(props, getSharedEventState);
+      const dataKeys = Object.keys(baseProps).filter((key) => key !== "parent");
```

Comment:
You're importing lodash `_.keys` elsewhere. Perhaps decide on one canonical way to do it? (Unless there's some reason for `_.keys` above that `Object.keys` can't do).